reference/src/glossary.md

```diff
@@ -146,6 +146,30 @@ Moreover, such unsafe code must not return a non-UTF-8 string to the "outside" o
 To summarize: *Data must always be valid, but it only must be safe in safe code.*
 For some more information, see [this blog post](https://www.ralfj.de/blog/2018/08/22/two-kinds-of-invariants.html).
 
+#### Undefined Behavior
+[ub]: #undefined-behavior
+
+*Undefined Behavior* is a concept of the contract between the Rust programmer and the compiler:
+The programmer promises that the code exhibits no undefined behavior.
+In return, the compiler promises to compile the code in a way that the final program does on the real hardware what the source program does according to the Rust Abstract Machine.
+If it turns out the program *does* have undefined behavior, the contract is void, and the program produced by the compiler is essentially garbage (in particular, it is not bound by any specification; the program does not even have to be well-formed executable code).
```

Comment:
Where are the nasal demons? :smile:

Reply:
Should we go as fast as saying that: 

> If it turns out the program *does* have undefined behavior, the contract is void, and the program is not a Rust program.

? 

That kind of clarifies that we are in the business of defining the behavior of Rust programs, and since those aren't Rust programs, we don't really need to say anything about them. Asking "what's their behavior" is as useful as asking "what's the behavior of trying to compile a C++ `.cpp` file with rustc.

Reply:
I meant that a reference to the nasal demons joke is missing :)

Reply:
(I know - sorry, my reply wasn't addressed at you, but at the snippet of text)

Reply:
I think "what is a Rust program" is getting into philosophy. Like, it is still syntactically a Rust program.

I think it is easier to try and carve-out the subset of "well-behaved" Rust programs than to exclude some things from the set of "Rust programs" all together.

Reply:
> I meant that a reference to the nasal demons joke is missing :)

Yeah, that was quite deliberate. ;)